[MRG + 1] Fix the cross_val_predict function for method='predict_proba'

dalmia opened the pull request:
#### Reference Issue
Fixes #7863 

#### What does this implement/fix? Explain your changes.
cross_val_predict() did not specify the order of the classes when method='predict_proba' was passed as the argument. So firstly, the change mentions in the documentation that the columns returned would be sorted. Secondly, for the estimators having a `classes_` attribute, it ensures that the predictions are returned in a sorted manner.

#### Any other comments?
Currently, I haven't yet tested the code. Would be completing that within a day or so.

dalmia committed:
Fix the cross_val_predict function for method='predict_proba'

dalmia committed:
Fix the syntax error and cleaning up code

dalmia committed:
STY: style fix (PEP8)

dalmia committed:
BUG: remove bug in _fit_and_predict

amueller commented:
``classes_`` is always sorted, so this sorting never does something.

dalmia commented:
@amueller I had the same doubt before I started working on the issue. However, the discussion in the thread mentioned that the classes are not always promised to be return in sorted order. Alsi, for classifiers like RandomForestClassifier, the docstring doesn't mention that the classes_ attribute return the classes in sorted manner. @jnothman Please provide your views too in this regard.

jnothman commented:
> classes_ is always sorted, so this sorting never does something.

Do we require that to be true for external classifiers?

amueller commented:
It's undocumented and untested, so I guess it's not required. We could require it. I'm not sure where in scikit-learn we might rely on it.

jnothman commented:
actually, it is tested:
https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/utils/estimator_checks.py#L1192

So I suppose we should worry about absent but not out-of-order classes...

On 17 November 2016 at 10:14, Andreas Mueller <notifications@github.com>
wrote:

> It's undocumented and untested, so I guess it's not required. We could
> require it. I'm not sure where in scikit-learn we might rely on it.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/7889#issuecomment-261103722>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz69gdKghBOtXG-GfmiPr1n1cYuWsMks5q-45ngaJpZM4Kz1fC>
> .
>

dalmia commented:
With this being tested, it's true that we don't need to worry about them being out-of-order. To check for absent classes though, checking that number of unique elements in `y` in `_fit_and_predict` equals the # columns of `predictions` works?

jnothman commented:
not necessarily, no. They may have differing subsets of classes, in a rare
case.

On 17 November 2016 at 17:58, Aman Dalmia <notifications@github.com> wrote:

> With this being tested, it's true that we don't need to worry about them
> being out-of-order. To check for absent classes though, checking that
> number of unique elements in y in _fit_and_predict equals the # columns
> of predictions works?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/7889#issuecomment-261170703>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz69tXkKilEkbVcD2ooAhSbVLTUuiKks5q-_sagaJpZM4Kz1fC>
> .
>

amueller commented:
I guess ``_fit_and_score`` needs to return ``classes`` and then we need to ensure compatibility / possibly pad the probabilities. There was a PR for making ``_fit_and_score`` return a dict, right? @raghavrv?

amueller commented:
This reminds me of a long-ago PR that added ``classes`` as an ``__init__`` argument for all estimators. Here we know all the classes in advance.... I'm wondering whether we couldn't do something more general and use a meta-estimator to make sure a classifier produces ``predict_proba`` and ``decision_function`` of the right shape.

jnothman commented:
Yes, I was trying to think of a more general solution... I don't mind a
meta-estimator, and it could be similarly applied to calibration where
we've just made similar fixes; but we have to be careful about backwards
compatibility of user interaction with the model.

On 18 November 2016 at 02:01, Andreas Mueller <notifications@github.com>
wrote:

> This reminds me of a long-ago PR that added classes as an __init__
> argument for all estimators. Here we know all the classes in advance....
> I'm wondering whether we couldn't do something more general and use a
> meta-estimator to make sure a classifier produces predict_proba and
> decision_function of the right shape.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/7889#issuecomment-261269473>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz61rg8fD1xCEnj0-ctx8VoOhMjN5gks5q_GxRgaJpZM4Kz1fC>
> .
>

amueller commented:
I was wondering whether we can create a meta-estimator without adding parameter indirection via ``estimator__``. ``set_params`` and ``get_params`` can be transparent, but ``__init__`` can not be, because ``**kwargs`` are not allowed (someone recently complained about that).

We could get around that using meta-classes, though I'm not sure if that'll look nice. I'm trying to come up with a better solution....

jnothman commented:
meta-estimator or not, this needs tests for:
* predict_proba, decision_function, predict_log_proba
* binary, multiclass, multilabel and multioutput problems (use a DecisionTreeClassifier, for instance)

jnothman commented:
i've marked it WIP: untested changes are not ready for merge

dalmia commented:
@jnothman @amueller Actually I am not familiar with what a meta-estimator is and hence, am unable to understand the solution that is being proposed. Could you people please give me a small explanation as to what is being discussed and how is a meta estimator different from a common estimator so that I can participate constructively as well? 
Thanks a lot.

jnothman commented:
A meta-estimator is just an estimator that wraps an estimator.

Regardless of whether we use a meta-estimator in the solution, the tests are the same.

@amueller, I've said before, you can make metaestimators without parameter indirection using polymorphic `clone` as discussed in #5080

amueller commented:
@jnothman You mean with polymorphic clone we could allow ``**kwargs`` in ``__init__``? Not sure that's my preferred solution.

amueller commented:
```python
from sklearn.preprocessing import LabelEncoder

class ClassesMeta(type):
    def __call__(cls, *args, **kwargs):
        classes = kwargs.pop("classes", None)
        obj = type.__call__(cls, *args, **kwargs)
        obj.classes = classes
        return obj

class ClassesMixin(object):
    def get_params(self, deep=True):
        params = super(ClassesMixin, self).get_params(deep=deep)
        params['classes'] = self.classes
        return params

    def set_params(self, params):
        self.classes = params.pop(classes, None)
        return super(ClassesMixin, self).set_params(params)
    
    def fit(self, X, y):
        self._private_le = LabelEncoder().fit(self.classes)
        return super(ClassesMixin, self).fit(X, self._private_le.transform(y))
    
    def predict(self, X):
        return self._private_le.inverse_transform(super(ClassesMixin, self).predict(X))
    
    def predict_proba(self, X):
        probs = super(ClassesMixin, self).predict_proba(X)
        padded_probs = np.zeros((probs.shape[0], len(self.classes)))
        class_mapping = np.searchsorted(self._private_le.classes_, self.classes_)
        padded_probs[:, class_mapping] = probs
        return padded_probs
        
def add_classes_wrapper(cls):
    return ClassesMeta(cls.__name__ + "WithClasses", (ClassesMixin, cls), {})
```
And then
```python
LogisticRegressionWithClasses = add_classes_wrapper(LogisticRegression)
asdf = LogisticRegressionWithClasses(classes=[0, 1, 2, 4])

from sklearn.datasets import load_iris
iris = load_iris()
asdf.fit(iris.data, iris.target)
asdf.predict_proba(iris.data).shape
```
> (150, 4)

Not the solution we want to use here probably though ;) Just wanted to see how bad it would be.
I'd say not that bad, actually.... It's a bit awkward that ``self.classes`` is not sorted though.

dalmia commented:
@amueller I am trying to work on this but feel really confused on how to start. Should I create a meta-estimator or run the tests mentioned by @jnothman separately? Please help me get started.

amueller commented:
You should check out my suggestion here:
https://github.com/scikit-learn/scikit-learn/pull/7889#issuecomment-261268374

And ignore everything afterwards.

dalmia commented:
But it seems that the `cross_val_predict` is independent of the `_fit_and_scores` function, so could you please explain your suggestion? Please correct me if I am mistaken.
Thanks.

dalmia commented:
Did you mean `_fit_and_predict`? (Since `_fit_and_scores` is used by `cross_val_score`)

dalmia committed:
FIX: making _fit_and_predict return classes

dalmia commented:
I made `_fit_and_predict` return classes. But the problem is that different cross validations might return slightly different sets of classes on a rare occasion and I'm trying to find a good solution for it. Please give your views on the issue @jnothman @amueller.

jnothman commented:
You would really benefit from writing a test case that currently fails, but you would like to succeed with your patch. That, especially if you then propose a solution, would much better help us decide the appropriate direction.

dalmia commented:
I did find a case where the current one fails with a slight modification of the iris dataset.
```python
iris = load_iris()
X, y = iris.data, iris.target
X, y = shuffle(X, y, random_state=0)
classes = len(set(y))
y[:50] = 0
y[50:100] = 1
X[:100], y[:100] = shuffle(X[:100], y[:100], random_state=0)
```
Now running `cross_val_predict` with `cv=kfold`:
```python
kfold = KFold(3)
predictions = cross_val_predict(est, X, y, method=method, cv=kfold)
```
This gives the following error as the training set has only 2 classes which reduces the dimensions of `predictions` to 1:
```python
ValueError                                Traceback (most recent call last)
<ipython-input-35-82b123df6101> in <module>()
     18 
     19     predictions = cross_val_predict(est, X, y, method=method,
---> 20                                     cv=kfold)
     21     assert_array_almost_equal(expected_predictions, predictions)

/home/aman/.local/lib/python2.7/site-packages/sklearn/model_selection/_validation.pyc in cross_val_predict(estimator, X, y, groups, cv, n_jobs, verbose, fit_params, pre_dispatch, method)
    416         predictions = sp.vstack(predictions, format=predictions[0].format)
    417     else:
--> 418         predictions = np.concatenate(predictions)
    419     return predictions[inv_test_indices]
    420 

ValueError: all the input arrays must have same number of dimensions
```
The solution I could think of was to use `LabelBinarizer` and fit it on `y` when the method is `predict_proba` in `_fit_and_predict` and transform the predictions each time before returning. Also, then we won't need to return `classes` in `_fit_and_predict` too. Seems it'll work?

dalmia committed:
DOC: Added docstring for the returned classes

amueller commented:
I forgot that ``_fit_and_predict`` gets the whole ``y``, so yeah, that seems to work.

dalmia commented:
I am getting the following error on running `nosetests` after compiling:
```python
Failure: ValueError (numpy.dtype has the wrong size, try recompiling. Expected 88, got 96) ... ERROR

======================================================================
ERROR: Failure: ValueError (numpy.dtype has the wrong size, try recompiling. Expected 88, got 96)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/nose/loader.py", line 414, in loadTestsFromName
    addr.filename, addr.module)
  File "/usr/lib/python2.7/dist-packages/nose/importer.py", line 47, in importFromPath
    return self.importFromDir(dir_path, fqname)
  File "/usr/lib/python2.7/dist-packages/nose/importer.py", line 94, in importFromDir
    mod = load_module(part_fqname, fh, filename, desc)
  File "/media/aman/BE66ECBA66EC7515/Open Source/scikit-learn/sklearn/__init__.py", line 57, in <module>
    from .base import clone
  File "/media/aman/BE66ECBA66EC7515/Open Source/scikit-learn/sklearn/base.py", line 12, in <module>
    from .utils.fixes import signature
  File "/media/aman/BE66ECBA66EC7515/Open Source/scikit-learn/sklearn/utils/__init__.py", line 10, in <module>
    from .murmurhash import murmurhash3_32
  File "__init__.pxd", line 155, in init sklearn.utils.murmurhash (sklearn/utils/murmurhash.c:6319)
ValueError: numpy.dtype has the wrong size, try recompiling. Expected 88, got 96

----------------------------------------------------------------------
Ran 1 test in 0.001s

FAILED (errors=1)
Makefile:32: recipe for target 'test-code' failed
make: *** [test-code] Error 1
```

Could you please help me understand the error here?

dalmia committed:
FIX: added LabelBinarizer to ensure predict_proba returns prdictions of the current shape

jnothman commented:
It should be easy to write an example where the current code fails, by passing a custom `cv` that creates pathological class distributions. Use such an example as a test case.

dalmia commented:
As I mentioned above, a variation of the `iris` dataset with `cv=kfold` for `n_splits=3`:
```python
y[:50] = 0
y[50:100] = 1
X[:100], y[:100] = shuffle(X[:100], y[:100], random_state=0)
```
Which gives the train labels in one of the folds as composed of only 2 classes. This breaks in `cross_val_predict` as expected.
However, my earlier proposed solution won't work as `LabelBinarizer` can't handle multioutput `y` (for method=`predict_proba`). I have thought of a workaround. This may not look very neat but it works. Please review.

dalmia committed:
FIX: fixed _fit_and_predict and added tests

dalmia commented:
As I mentioned above, a variation of the `iris` dataset with `cv=kfold` for `n_splits=3`:
```python
y[:50] = 0
y[50:100] = 1
X[:100], y[:100] = shuffle(X[:100], y[:100], random_state=0)
```
Which gives the train labels in one of the folds as composed of only 2 classes. This breaks in `cross_val_predict` as expected.
However, my earlier proposed solution won't work as `LabelBinarizer` can't handle multioutput `y` (for method=`predict_proba`). I have thought of a workaround. This may not look very neat but it works. Please review.

jnothman commented on sklearn/model_selection/tests/test_validation.py:
```diff
@@ -914,6 +914,10 @@ def test_cross_val_predict_with_method():
     X, y = shuffle(X, y, random_state=0)
     classes = len(set(y))
 
+    y[:50] = 0
```
It would be more normal to make a separate test for tricky cases, rather than overload the test for basic functionality. This helps diagnose, when a test fails, what went wrong.

Certainly, this dataset modification needs a comment as to its purpose.

jnothman commented on sklearn/model_selection/_validation.py:
```diff
@@ -474,6 +476,18 @@ def _fit_and_predict(estimator, X, y, train, test, verbose, fit_params,
         estimator.fit(X_train, y_train, **fit_params)
     func = getattr(estimator, method)
     predictions = func(X_test)
+    if method in ['decision_function', 'predict_proba', 'predict_log_proba']:
```
I wonder if we should be inspecting the shape of the prediction, rather than doing this based on name.

dalmia replied:
The problem lies is that `predict_proba` and `predict_log_proba` return a 2D array when there are 2 classes in the training set but `decision_function` returns a 1D array. So, I chose to go by method to differentiate the two.

jnothman commented on sklearn/model_selection/_validation.py:
```diff
@@ -474,6 +476,18 @@ def _fit_and_predict(estimator, X, y, train, test, verbose, fit_params,
     predictions = func(X_test)
+    if method in ['decision_function', 'predict_proba', 'predict_log_proba']:
+        true_classes = np.unique(y)
+        train_classes = np.unique(y_train)
```
I think we should be using the estimator's `classes_` here

dalmia replied:
But a few estimators implementing the `predict_proba` method, don't seem to have `classes_` as an attribute like `SVC`.

jnothman replied:
Huh?
```python
>>> SVC().fit(np.arange(20).reshape(5, 4), [0, 1, 1, 0, 1]).classes_
array([0, 1])
```

dalmia replied:
Sorry, I just looked at the [documentation](http://scikit-learn.org/stable/modules/generated/sklearn.svm.SVC.html) and made the note. I'll add the change.

jnothman commented on sklearn/model_selection/_validation.py:
```diff
@@ -474,6 +476,18 @@ def _fit_and_predict(estimator, X, y, train, test, verbose, fit_params,
+        train_classes = np.unique(y_train)
+        predictions_ = np.zeros((X_test.shape[0], true_classes.shape[0]))
+        if method is 'decision_function' and len(train_classes) == 2:
+            class_predictions = estimator.predict(X_test)
```
I don't see what the output of `predict` has to do with it?

dalmia replied:
I realize I made a mistake here. Will correct this.

jnothman commented on sklearn/model_selection/_validation.py:
```diff
@@ -474,6 +476,18 @@ def _fit_and_predict(estimator, X, y, train, test, verbose, fit_params,
+            for i, j in enumerate(class_predictions):
+                predictions_[i, j] = predictions[i]
+        else:
+            for i, j in enumerate(train_classes):
```
I don't see how this is mapping into the `true_classes` space

dalmia committed:
TST: added separate test case and removed  from tests

dalmia commented:
I added another test case as you mentioned. But, I have removed `decision_function` from the changes that I have made. This is because, `decision_function` produces different shapes based on whether `decision_function_shape` is `ovo` or `ovr`. Now, for the case where say, the true classes are `[0,1,2,3]` but a particular split only has classes `[0,1]` in the training set, this makes the implementation of my solution for `predict_proba` and `log_predict_proba`, a bit clumsy when applied to `decision_function`. So, I was thinking of adding a separate check for the method being `decision_function`. Please give your suggestions on this.

jnothman commented:
Please add tests where:
* the classes are represented as strings
* the classes are represented as ints, but not ranging from 0 (e.g. starting at 1, or including negative numbers)
* `decision_function` is a problem

Then solve these.

You may wish to use a `LabelEncoder`. You may wish to test for the shape of each method call's return value only after getting all of them back. There are a few ways to spin this, but I'd like to see a somewhat more general solution.

jnothman commented on sklearn/model_selection/_validation.py:
```diff
@@ -474,6 +476,13 @@ def _fit_and_predict(estimator, X, y, train, test, verbose, fit_params,
+        true_classes = np.unique(y)
+        train_classes = np.unique(y_train)
+        predictions_ = np.zeros((X_test.shape[0], true_classes.shape[0]))
+        for i, j in enumerate(train_classes):
```
Once you've encoded the classes outside of this function, I think this loop can be performed as `predictions_[:, estimator.classes_] =  predictions` and in the case of decision_function with binary problems, `predictions_[:, estimator.classes_[-1]] =  predictions`.

dalmia replied:
This makes it so much cleaner. Thanks @jnothman

dalmia committed:
FIX: improved _fit_and_predict for class labels

dalmia commented:
@jnothman Thanks for the suggestions. They really helped me proceed. I implemented the support as you mentioned. However, I assumed you wanted me to add the tests for classes as strings just to debug and arrive at the correct solution. Did you intend to say that I should add the tests in `test_validation.py`?

amueller commented:
yes, please add a test for the problem to the test suite.

dalmia commented:
@amueller Yes, I'll do it. Also, flake8 errors are showing up in files that I did not change. I don't suppose this is the natural behaviour?

dalmia committed:
TST: added tests for different label types

dalmia committed:
FIX: resolved errors

dalmia commented:
I guess there is some problem with the build. `nosetests` pass when run locally, but show error in AppVeyor. Also, Travis is showing flake8 errors on files in `master` which are not present in my branch. Please look into this.

maniteja123 commented:
Hi, looking at the [logs](https://travis-ci.org/scikit-learn/scikit-learn/jobs/185103460#L484) it seems it is picking up the diff between master and your branch. I suppose rebasing on master will solve the flake8 issue. Regarding the appveyor tests, [error](https://ci.appveyor.com/project/sklearn-ci/scikit-learn/build/1.0.10977/job/dq6c5k1p506au53h#L10905) AFAICS the tests are failing because of incorrect indexing. I think the value of ``est.classes_`` is ``[1, 2, 3]`` and 3 is an incorrect index for axis = 1 which is size 3. I don't entirely know the use case here but hope it helps !

dalmia commented:
@maniteja123 Thanks for the tip on travis, but still I think this shouldn't be showing up the error when running on my branch. Also, I did test it and the value of `est.classes_` is `[0,1,2]` when am running on my device. That's why I thought that there's a problem with the builds.

maniteja123 commented:
Hi, sorry I didn't try to run the code before commenting. Actually looking at it, the tests seem to succeed for other appveyor and travis tests. Perhaps you can add a debug statement and push a commit to investigate further. Regarding flake8, I am not sure I know the reason for the error but flake8 does throw those errors on current master. I have just checked it. Please do let me know if you can also reproduce them.

jnothman reviewed (commented):
sorry this is only a partial review. need to look through tests. I am surprised that some of those flake8 failures are showing up here seeing as they're on unrelated files.

dalmia commented:
@maniteja123 Thanks for looking up. You are right, that flake8 does throw those errors on master. I think rebasing on master seems the viable thing to do.

dalmia commented on sklearn/model_selection/tests/test_validation.py:
```diff
@@ -936,6 +937,124 @@ def test_cross_val_predict_with_method():
         assert_array_almost_equal(expected_predictions, predictions)
 
 
+def test_cross_val_predict_corner_case():
```
Seems better. Also, I just thought of something. Since this is a test for a rare edge case, how about `test_cross_val_predict_edge_case`?

jnothman replied:
Huh? How is that more descriptive than "corner case"? Why would we prefer it over "class subset"?

dalmia replied:
Although we are modifying a subset of our samples, still somehow I am unable to relate 'class subset' to this test representing a rare case. Over 'corner case' though, 'edge case' seemed better in brevity.

jnothman replied:
No. There's not much benefit in extreme brevity for test names. On the other hand "edge case" is highly undescriptive of which edge case when there are no doubt many.

dalmia replied:
@jnothman Yes, I understand it now from a moderator point of view. Thank you for making it clear to me. I'll make the change.

dalmia committed:
FIX: replaced `is` with `==`  for comparison

dalmia committed:
FIX: update test name

jnothman commented on sklearn/model_selection/tests/test_validation.py:
```diff
@@ -936,6 +937,138 @@ def test_cross_val_predict_with_method():
+    # kfold.split, the training set is composed of only 2 classes instead of 3
+    y[:50] = 0
+    y[50:100] = 1
+    X[:100], y[:100] = shuffle(X[:100], y[:100], random_state=0)
```
Could we have just done something like
```python
X = np.arange(6).reshape(3, 2)
y = [1,2,3]
cross_val_predict(LogisticRegression(), X, y, cv=3, method=method)
```
to get the same effect, i.e. broken in master and fixed here?

Even better, use
```python
X = np.arange(8).reshape(4, 2)
y = [1,1,2,3]
```
with `KFold(3)` and `KFold(4)` to create two different kinds of broken behaviour in master.

jnothman replied:
(I've corrected cv=3 to cv=KFold(3))

To be sure, what you've got here looks like it tests the issue, but it's not easy to see that it does so.

dalmia replied:
I agree that the tests might not be that obvious in testing the issue, but they did fail on the cases we needed them to and worked in my implementation. The ones you suggested seem much cleaner. I'll make the change.

jnothman commented on sklearn/model_selection/tests/test_validation.py:
```diff
@@ -936,6 +937,138 @@ def test_cross_val_predict_with_method():
+        func = getattr(est, method)
+
+        # Naive loop (should be same as cross_val_predict):
+        for train, test in kfold.split(X, y):
```
I don't think this is as clear as just checking whether the mask `predictions == 0` matches precisely those elements where the classifier has no information.

But that might be a matter of taste, I'll admit.

dalmia replied:
Oh yes. This could have been much simpler to implement in the first place. I can change it if it's absolutely necessary?

dalmia replied:
Yes, I'll give it some more thought once the basic tests are done.

jnothman commented on sklearn/model_selection/tests/test_validation.py:
```diff
@@ -936,6 +937,138 @@ def test_cross_val_predict_with_method():
+        y_str[np.where(y == i)] = c
+
+    # Modifies the dataset so that in a particular fold, produced by
+    # kfold.split, the training set is composed of only 2 classes instead of 3
```
Do we need this edge case here?

jnothman replied:
Is it not appropriate to just check in `test_cross_val_predict_with_method` whether alternative representations of y (including `y + 1`, `y - 2`, `y.astype('U')`) yield the same predictions?

dalmia replied:
Yes indeed. Thanks.

dalmia commented:
The rest of the things work as expected, but there was one line that broke:
```python
predictions = cross_val_predict(est, X, y, method=method)
```
When we don't pass any value for `cv`, I get this error which otherwise doesn't appear for `cv=KFold(3)` or `cv=KFold(4)`:
```python
ValueError                                Traceback (most recent call last)
<ipython-input-5-5946aa69faaa> in <module>()
----> 1 test_cross_val_predict_class_subset()

<ipython-input-4-aa59a4580ab2> in test_cross_val_predict_class_subset()
     14         est = LogisticRegression()
     15 
---> 16         predictions = cross_val_predict(est, X, y, method=method)
     17         assert_equal(len(predictions), len(y))
     18 

/media/aman/BE66ECBA66EC7515/Open Source/scikit-learn/sklearn/model_selection/_validation.py in cross_val_predict(estimator, X, y, groups, cv, n_jobs, verbose, fit_params, pre_dispatch, method)
    389 
    390     cv = check_cv(cv, y, classifier=is_classifier(estimator))
--> 391     cv_i = list(cv.split(X,y))
    392     cv_iter = list(cv.split(X, y, groups))
    393 

/media/aman/BE66ECBA66EC7515/Open Source/scikit-learn/sklearn/model_selection/_split.pyc in split(self, X, y, groups)
    320                                                              n_samples))
    321 
--> 322         for train, test in super(_BaseKFold, self).split(X, y, groups):
    323             yield train, test
    324 

/media/aman/BE66ECBA66EC7515/Open Source/scikit-learn/sklearn/model_selection/_split.pyc in split(self, X, y, groups)
     89         X, y, groups = indexable(X, y, groups)
     90         indices = np.arange(_num_samples(X))
---> 91         for test_index in self._iter_test_masks(X, y, groups):
     92             train_index = indices[np.logical_not(test_index)]
     93             test_index = indices[test_index]

/media/aman/BE66ECBA66EC7515/Open Source/scikit-learn/sklearn/model_selection/_split.pyc in _iter_test_masks(self, X, y, groups)
    611 
    612     def _iter_test_masks(self, X, y=None, groups=None):
--> 613         test_folds = self._make_test_folds(X, y)
    614         for i in range(self.n_splits):
    615             yield test_folds == i

/media/aman/BE66ECBA66EC7515/Open Source/scikit-learn/sklearn/model_selection/_split.pyc in _make_test_folds(self, X, y, groups)
    576             raise ValueError("All the n_groups for individual classes"
    577                              " are less than n_splits=%d."
--> 578                              % (self.n_splits))
    579         if self.n_splits > min_groups:
    580             warnings.warn(("The least populated class in y has only %d"

ValueError: All the n_groups for individual classes are less than n_splits=3.

In [4]:
```

I could deduce that the `cv` that `check_cv` chose was `StratifiedKFold`. Any reason this should be appearing?

dalmia committed:
TST: improved the tests

dalmia commented:
Removed the line mentioned above from the tests for now.

jnothman commented:
Yes, our StratifiedKFold won't work with so few instances.

jnothman commented:
But there's no reason to test with it in this context.

dalmia commented:
Okay, so please review my changes.

jnothman commented on sklearn/model_selection/tests/test_validation.py:
```diff
@@ -935,6 +936,95 @@ def test_cross_val_predict_with_method():
+        predictions = cross_val_predict(est, X, y, method=method,
+                                        cv=kfold3)
+
+        # Naive loop (should be same as cross_val_predict):
```
To avoid repeating code (if we're keeping this rather than just checking the presence of zeros/nonzeros) I'd rather you pulled this loop and assertion out as a function.

dalmia replied:
Yes, I'll do that.

jnothman reviewed (changes requested):
Otherwise LGTM. Except that I'd still rather avoid having to use hard-coded method names in the implementation if that can be avoided.

dalmia commented:
About hardcoding `decision_function`, I'll still try to think of some workaround but it's highly unlikely as only `decision_function` in the binary case has to be tackled separately.

dalmia committed:
ENH: added function to expected predictions

dalmia committed:
FIX: remove flake8 errors

dalmia commented:
@raghavrv Thanks for letting me know. Yes, I'll look into it.

dalmia committed:
FIX: replace unicode conversion with str conversion

dalmia commented:
Is this ready for merge?

raghavrv commented on sklearn/model_selection/_validation.py:
```diff
@@ -474,6 +481,14 @@ def _fit_and_predict(estimator, X, y, train, test, verbose, fit_params,
     func = getattr(estimator, method)
     predictions = func(X_test)
+    if method in ['decision_function', 'predict_proba', 'predict_log_proba']:
+        true_classes = np.unique(y)
```
Why not have `_, n_classes =  np.unique(y, return_counts=True)` as you don't use the "true_classes"?

dalmia replied:
As mentioned by @jnothman that `return_counts` is not supported for all versions and it's true that only the number of classes is being used, I have thought of another workaround. Will do the change.

raghavrv commented on sklearn/model_selection/tests/test_validation.py:
```diff
@@ -935,6 +936,79 @@ def test_cross_val_predict_with_method():
+        assert_array_almost_equal(expected_predictions, predictions)
+
+        # Testing unordered labels
+        y = [1, 1, -4, 6]
```
Do you want to explicitly test for a use-case where the `classes_` is guaranteed to not be sorted but `cross_val_predict` returns output in sorted order?

dalmia replied:
As mentioned [here](https://github.com/scikit-learn/scikit-learn/pull/7889#issuecomment-261149076), `classes_` should indeed be sorted. Please have a look.

raghavrv replied:
Fair enough! Thanks for checking :)

raghavrv replied:
But wait could we have a mock estimator that does not have the `classes_` sorted? @jnothman Is it worth having such a test?

raghavrv replied:
Never mind this has been discussed before. Sorry for not checking... +1 for merge with the whatsnew entry...

dalmia committed:
FIX: replaced 'true_classes' with 'n_classes

raghavrv commented:
And a whatnew entry please...

dalmia committed:
DOC: added whats_new entry

dalmia commented:
I am getting this error also when I am building locally:
```python
Exception occurred:
  File "/home/ubuntu/scikit-learn/doc/sphinxext/sphinx_gallery/docs_resolv.py", line 48, in _get_data
    with open(url, 'r') as fid:
IOError: [Errno 2] No such file or directory: '/home/ubuntu/scikit-learn/doc/_build/html/stable/modules/generated/sklearn.feature_selection.SelectKBest.rst.html'
The full traceback has been saved in /tmp/sphinx-err-wz9PtY.log, if you want to report the issue to the developers.
Please also report this if it was a user error, so that a better error message can be provided next time.
A bug report can be filed in the tracker at <https://github.com/sphinx-doc/sphinx/issues>. Thanks!
Embedding documentation hyperlinks in examples..
	processing: feature_stacker.html
make: *** [html] Error 1
```
Any workaround for this?

jnothman commented:
Merging with an updated master will fix it for CircleCI

raghavrv commented:
+1 for merge after updating master (and all CIs turning green)

jnothman commented:
@raghavrv, nothing here could cause the doc build to fail.

jnothman commented:
Merged, thanks @dalmia!

AlJohri commented:
I started getting a different shape for `cross_val_predict` with the method of `decision_function` for this model: `svm.SVC(kernel='linear', C=1, probability=False)` with this `cross_val_predict(original_pipeline, X_train, y_train, cv=2, method='decision_function')`. I was getting a 1D array before and now I'm getting a 2D array where the first column is all zeroes. I'm doing a binary classifier.
